apps/astarte_data_updater_plant/lib/astarte_data_updater_plant/rpc/impl.ex

```diff
@@ -0,0 +1,172 @@
+defmodule Astarte.DataUpdaterPlant.RPC.Impl do
+  alias ElixirSense.Log
+  alias Astarte.DataUpdaterPlant.RPC.State
+  alias Astarte.DataUpdaterPlant.RPC.Device
+  alias Astarte.DataUpdaterPlant.RPC.Queries
+  require Logger
+
+  # State initialization by retrieving state via DB queries
+  def init_state() do
+    Logger.info("Initializing dup rpc handler state ...")
+    Logger.info("Fetching realms ...")
+
+    realms = retrieve_realms()
+    Logger.info("Retrieved realms: #{inspect(realms)}")
+
+    devices =
+      realms
+      |> Enum.reduce(%{}, fn realm, acc ->
+        Logger.info("Fetching connected devices for realm: #{realm} ...")
+
+        case retrieve_connected_devices(realm) do
+          {:ok, realm_devices} ->
+            Logger.info("Connected devices for realm #{realm} are: #{inspect(realm_devices)}")
+
+            Enum.reduce(realm_devices, acc, fn device, acc_map ->
+              Map.put(acc_map, device.device_id, device)
+            end)
+
+          {:error, reason} ->
+            Logger.error("Failed to fetch devices for realm #{realm}: #{inspect(reason)}")
+            acc
+        end
+      end)
+
+    %State{
+      devices: devices
+    }
```

Comment:
It seems the code is populating a single `devices` map with all the devices of all the realms, using `device_id` as key and `device` as value.
However, multiple realms can have the same `device_id` registered, which still represent two different devices.
So the code should be updated, otherwise the final map will list only one device ID / device, even though there may be multiple devices with the same device ID (just on different realms).

You could evaluate to use a `{realm_name, device_id}` key for the map.